### DISCUSSION TITLE
Is low memory operation in scope for gitoxide?

### DISCUSSION THREAD

**Discussion by tanriol** [GitoxideLabs/gitoxide#1699] [Q&A]
Recently I wanted to mirror several big git reporisories by cloning them (as bare) and then periodically fetching updates on a low-end NAS. Turned out neither git nor gitoxide can currently be used to clone something like https://git.kernel.org/pub/scm/linux/kernel/git/stable/linux.git/ while having just 512MB RAM.

I've failed to find whether being usable in low-memory environments is a goal for gitoxide. Would it make sense to try adding support for such use cases or would trying to support these without degrading performance in the normal case be unwanted complexity for `gitoxide` that'd be really unlikely to be accepted?

**Comment by Byron** [comment]
Thanks for sharing, that's a very interesting comparison and I am somewhat glad that `gitoxide` isn't worse than Git here.

However, it's also not a project goal to support smaller systems, particularly not 32bit systems which will have more trouble than with Git. Most 'small' systems nowadays are 64 bit, so the greatest obstacle towards completing a task will be swapping.

When you say that cloning the linux kernel with 512MB RAM didn't work, what does that mean exactly? OOM kill? Or excruciating swapping? If the case is the latter, I would have expected it to finish eventually.

Normally, I'd say that there is no way that this is made to work with the current algorithm - it's already mindful about memory and should be a bit better than Git, but it still has to hold millions of objects in memory all at once, which is where the high demands for the linux kernel are coming from.

However, there is [an issue whose resolution](https://github.com/GitoxideLabs/gitoxide/issues/1025) basically requires rewriting the way packs are resolved. This algorithm will be more compatible, but also will be less parallel. While at it, I don't think it would hurt to make it possible to work in low-memory situations. To make that work, the only thing I can imagine that it crates some on-disk memory-mappable data-structure that first has to grow while streaming the pack, to later be traversed. Interestingly I'd think that if the disk is moderately fast, that streaming of data shouldn't be a bottleneck neither for reading nor for writing, if done cleverly, (or if a sqlite database is used ;)).

That's a long-winded way of saying: A fix would be welcome if it would also fix #1025 at the same time.

**Reply by tanriol** [reply]
> When you say that cloning the linux kernel with 512MB RAM didn't work, what does that mean exactly? OOM kill? Or excruciating swapping? If the case is the latter, I would have expected it to finish eventually.

So far I've seen two problems with cloning big repos. The first appeared while cloning https://github.com/gentoo/gentoo.git (5M objects). When (towards the end of cloning) RAM becomes tight, realtime updates of the in-memory index start swapping and thus slowing down the clone operation to the point where (I think) github just cuts it (manifested with `gix` producing an IO error like "failed to fill whole buffer" IIRC). I've seemingly succeeded at doing that once with 32-bit build, possibly a matter of luck.
`git` in this case had some other network-looking problem.

The second happens with cloning https://git.kernel.org/pub/scm/linux/kernel/git/stable/linux.git/ and is different
`git`: after 24 hours or so of swapping was at ~20% of "resolving deltas", interrupted manually
`gix`: panicked after ~20 min (around 25% objects received) with "Unable to allocate another span" (looks like coming from something `tracing`-related, not reproducing on a desktop)

> [...]
> That's a long-winded way of saying: A fix would be welcome if it would also fix #1025 at the same time.

Okay, I'll look into that, thank you.

**Reply by Byron** [reply]
Thank you for elaborating, it's very interesting to hear how low-memory situations are 'expressing' themselves. From what I hear, even if the remote wouldn't cut the connection after a day, the initial clone is still too slow to happen.

Oh, before I forget. Something simpler than writing a new pack resolution algorithm could be to download bundles. That way, the initial download is just an existing pack file ~~*along with its index*~~… never mind, [bundles](https://git-scm.com/docs/git-bundle) are also just pack files, the index will have to be produced locally which is what's causing the problem in the first place.

So indeed, there seems to be no way around creating a low-memory way of resolving a pack if this should have a chance to work, a prospect I find very interesting to say the least. If you were to succeed, then `gitoxide` already served its purpose of *also* being a platform to experiment with novel approaches, something that should be easier in Rust to do for *more* people.

**Reply by NobodyXu** [reply]
Maybe using a different allocator could help?

I.e. mimalloc or jemalloc which has less fragmentation.

If it's swapping a lot...then maybe 
 - use zswap/zram which compresses the memory instead of swapping
 - increase the swapping partition
 - use a faster SSD

😅 just my two cents on how it could be speedup, it could be totally wrong for your environment

**Reply by Byron** [reply]
Thanks a lot for sharing!

It would indeed be interesting if such optimizations could already help with the completion of the operation today.

By the way, I noticed some sort of regression when handling the linux repository: on my machine, it uses 3.3GB of real memory, even though Git only goes up to 900MB or so. `gix` starts out at 2.3GB memory usage, something that might be explainable with the in-memory data structure that it builds, but goes up to 3.3GB during operation for unknown reasons - it reuses existing buffers so memory fragmentation shouldn't be the issue. Also, pack-verification isn't the same cloning, but similar enough.
Git on the other hand, stays at around 900MB at all times.

All this would make `gitoxide` less suitable on memory constrained systems right now. And I say this with some surprise as when I wrote this, I measured 20% less memory consumption.

When cloning, I measured even higher memory requirements at 5.2GB for `gix`, but `git` also went up to 5.5GB here. This leads me to believe that pack-verification in Git might work differently than it does for `gix`. But even running the `--algorithm less-memory` version starts out with 2.3GB and runs up to 3.3GB, just like the `less-time` version (the default).

```
/usr/bin/time -lp gix free pack verify -a less-memory $PWD/tests/fixtures/repos/linux.git/objects/pack/pack-b73c0dd365b8ab881a163a1d342dac7416f52591.idx
```

Definitely something to investigate at some point, even though in the end the real difference is in how pack-verification works - cloning seems similar after all.

<details><summary>clone</summary>

```
❯ /usr/bin/time -lp gix clone --bare linux linux-gix
 08:19:49 indexing done 10.5M objects in 37.40s (279.5K objects/s)
 08:19:49 decompressing done 6.9GB in 37.40s (185.8MB/s)
 08:19:55     Resolving done 10.5M objects in 6.08s (1.7M objects/s)
 08:19:55      Decoding done 139.6GB in 6.08s (23.0GB/s)
 08:19:57 writing index file done 322.3MB in 0.47s (681.1MB/s)
 08:19:57  create index file done 10.5M objects in 44.96s (232.5K objects/s)
 08:19:57          read pack done 3.0GB in 64.15s (46.7MB/s)
real 64.74
user 158.68
sys 9.64
          5270519808  maximum resident set size
                   0  average shared memory size
                   0  average unshared data size
                   0  average unshared stack size
              558879  page reclaims
              239740  page faults
                   0  swaps
                   0  block input operations
                   0  block output operations
                   0  messages sent
                   0  messages received
                  17  signals received
              124885  voluntary context switches
              419010  involuntary context switches
        746515556178  instructions retired
        425506347028  cycles elapsed
          1375572912  peak memory footprint
```

```
❯ /usr/bin/time -lp git clone --bare file://$PWD/linux linux-git
Cloning into bare repository 'linux-git'...
remote: Enumerating objects: 10452599, done.
remote: Counting objects: 100% (1095735/1095735), done.
remote: Compressing objects: 100% (259255/259255), done.
Receiving objects: 100% (10452599/10452599), 2.79 GiB | 85.42 MiB/s, done.
remote: Total 10452599 (delta 893263), reused 1003093 (delta 833868), pack-reused 9356864
Resolving deltas: 100% (8591842/8591842), done.
real 105.64
user 168.88
sys 115.80
          5504090112  maximum resident set size
                   0  average shared memory size
                   0  average unshared data size
                   0  average unshared stack size
              514650  page reclaims
                7186  page faults
                   0  swaps
                   0  block input operations
                   0  block output operations
                   0  messages sent
                   0  messages received
                 100  signals received
              922008  voluntary context switches
            13913826  involuntary context switches
          7398684272  instructions retired
          5698143955  cycles elapsed
             5407680  peak memory footprint
```

</details>

<details><summary>pack verify</summary>

```
❯ /usr/bin/time -lp git -c pack.threads=1 verify-pack $PWD/tests/fixtures/repos/linux.git/objects/pacreal 72.29c0dd365b8ab881a163a1d342dac7416f52591.idx
user 64.00
sys 8.11
           934739968  maximum resident set size
                   0  average shared memory size
                   0  average unshared data size
                   0  average unshared stack size
               59763  page reclaims
                 278  page faults
                   0  swaps
                   0  block input operations
                   0  block output operations
                   0  messages sent
                   0  messages received
                   0  signals received
                 500  voluntary context switches
                8254  involuntary context switches
            95133799  instructions retired
            43875330  cycles elapsed
             2146688  peak memory footprint
```

```
❯ /usr/bin/time -lp gix -t1 free pack verify $PWD/tests/fixtures/repos/linux.git/objects/pack/pack-b73c0dd365b8ab881a163a1d342dac7416f52591.idx
 08:13:24 Hash of index 'pack-b73c0dd365b8ab881a163a1d342dac7416f52591.idx' done 209.7MB in 0.11s (1.9GB/s)
 08:13:24                                           collecting sorted index done 7.5M entries in 0.45s (16.7M entries/s)
 08:13:24 Hash of pack 'pack-b73c0dd365b8ab881a163a1d342dac7416f52591.pack' done 1.3GB in 0.59s (2.2GB/s)
 08:13:25                                                          indexing done 7.5M objects in 0.93s (8.1M objects/s)5M objects [98%] [=============================================================>--]
 08:14:15                                                         Resolving done 7.5M objects in 49.81s (150.4K objects/s) objects [99%] |62.5K/s| [===================================================>-]
 08:14:15                                                          Decoding done 94.3GB in 49.81s (1.9GB/s)
real 51.40
user 49.18
sys 2.76
          3240165376  maximum resident set size
                   0  average shared memory size
                   0  average unshared data size
                   0  average unshared stack size
              198534  page reclaims
                 564  page faults
                   0  swaps
                   0  block input operations
                   0  block output operations
                   0  messages sent
                   0  messages received
                   0  signals received
                  83  voluntary context switches
                7078  involuntary context switches
        348378934968  instructions retired
        202708440801  cycles elapsed
          1194543096  peak memory footprint
```

</details>

**Reply by NobodyXu** [reply]
> By the way, I noticed some sort of regression when handling the linux repository: on my machine, it uses 3.3GB of real memory, even though Git only goes up to 900MB or so.

Could it be that the buffer allocated is larger than the actual data?

I.e. Even `Vec::reserve` grows the `Vec` exponentially and it would never free it up unless you `Vec::shrint_to*`.

Same for `HashMap`, maybe at some point it's over-reserved?

**Reply by Byron** [reply]
That's a great hint! 

Even though I don't think `reserve()` is used incorrectly (i.e. used as if it was absolute, even though it's relative, i.e. additional items), I definitely found plenty of instances where potentially large vectors are instantiated with `with_capacity()`, which will typically over-allocate. What's meant in this case is `reserve_exact()`.

I will change it and see if this has a noticeable impact, and wouldn't be surprised if it does.
Will post here once I have the data.

**Reply by tanriol** [reply]
> Maybe using a different allocator could help?

In some cases maybe. In the `gix` slowdown and connection cutoff case - not really, at that time most memory consumption happens in just two arrays (64 bytes per entry on 64-bit, 52 I think on 32-bit) that are accessed by binary search in a mostly random pattern I think.

> use a faster SSD

In my case it's a NAS without SSD slots, it's swapping to HDD, so swap-in has really high latency there.

Going to look into the RAM consumption and access patterns of these. Looks to me like it should be possible to change access patterns even without reducing peak RAM, but not certain yet.

**Reply by tanriol** [reply]
> Even though I don't think `reserve()` is used incorrectly (i.e. used as if it was absolute, even though it's relative, i.e. additional items), I definitely found plenty of instances where potentially large vectors are instantiated with `with_capacity()`, which will typically over-allocate.

I think that may be actually fine - it'd increase address space consumption, but not RAM consumption as untouched pages won't ever be allocated. Well, unless you have overcommit disabled.

**Reply by NobodyXu** [reply]
> In my case it's a NAS without SSD slots, it's swapping to HDD, so swap-in has really high latency there.

@tanriol I think you could try zswap/zram, it's designed for situations like this.

It will compress the page using whatever algorithm you pick instead of immediately swapping it out to a disk, and that could be beneficial if gix memory have good compression ratio, and only swap out if compressing the page isn't enough.

**Reply by Byron** [reply]
Right, I keep forgetting that modern OS's are very capable!

[The PR](https://github.com/GitoxideLabs/gitoxide/pull/1700) here experiments with this, and comes up with nothing, no change at all. To add insult to injury, I am unable to profile memory consumption on MacOS due to code-signature requirements that I apparently can't fulfil.

Maybe you could share the tooling you used so I could run that in a Linux VM?
